nexus/src/db/datastore.rs

```diff
@@ -238,6 +242,173 @@ impl DataStore {
+                    .filter(region_dsl::disk_id.eq(disk_id))
+                    .inner_join(
+                        dataset_dsl::dataset
+                            .on(region_dsl::dataset_id.eq(dataset_dsl::id)),
+                    )
+                    .select((Dataset::as_select(), Region::as_select()))
+                    .get_results::<(Dataset, Region)>(conn)?;
+                if !datasets_and_regions.is_empty() {
+                    return Ok(datasets_and_regions);
+                }
+
+                let datasets: Vec<Dataset> = dataset_dsl::dataset
+                    // We look for valid datasets (non-deleted crucible datasets).
+                    .filter(dataset_dsl::time_deleted.is_null())
+                    .filter(dataset_dsl::kind.eq(DatasetKind(
+                        crate::internal_api::params::DatasetKind::Crucible,
+                    )))
+                    // Next, observe all the regions allocated to each dataset, and
+                    // determine how much space they're using.
+                    //
+                    // TODO: We could store "free/allocated" space per-dataset,
+                    // and keep them up-to-date, rather than trying to recompute
+                    // this.
+                    //
+                    // TODO: We admittedly don't actually *fail* any request for
+                    // running out of space - we try to send the request down to
+                    // crucible agents, and expect them to fail on our behalf in
+                    // out-of-storage conditions. This should undoubtedly be
+                    // handled more explicitly.
+                    .left_outer_join(
+                        region_dsl::region
+                            .on(dataset_dsl::id.eq(region_dsl::dataset_id)),
+                    )
+                    .group_by(dataset_dsl::id)
+                    .select(Dataset::as_select())
+                    .order(
+                        diesel::dsl::sum(
+                            region_dsl::extent_size * region_dsl::extent_count,
+                        )
+                        .asc(),
+                    )
+                    .get_results::<Dataset>(conn)?;
+
+                if datasets.len() < REGION_REDUNDANCY_THRESHOLD {
+                    return Err(TxnError::CustomError(
+                        RegionAllocateError::NotEnoughDatasets(datasets.len()),
+                    ));
+                }
+
+                // Create identical regions on each of the following datasets.
+                let source_datasets = &datasets[0..REGION_REDUNDANCY_THRESHOLD];
+                let regions: Vec<Region> = source_datasets
+                    .iter()
+                    .map(|dataset| {
+                        Region::new(
+                            dataset.id(),
+                            disk_id,
+                            params.block_size().try_into().unwrap(),
```

Comment:
Are these safe?  I think if we use ByteCount for types, we can do this conversion without risk of panic.  (See the comment in `impl From<&ByteCount> for i64`.)